### DISCUSSION TITLE
Call a component of a normal vector in a Postprocessor

### DISCUSSION THREAD

**Discussion by TLWise** [idaholab/moose#26940] [Q&A General]
I need to use the y-component of _normal in my Postprocessor. Can this be accomplished and if so how what is the syntax to call it for use?

**Comment by Ali1990dashti** [comment]
Which module of MOOSE are you using to calculate normals? If you are using PorousFlow the calculation is quite straightforward via `AuxKernels `and `AuxVariables`:
```
[AuxVariables]
  [n1]
    order = CONSTANT
    family = MONOMIAL
  []
  [n2]
    order = CONSTANT
    family = MONOMIAL
  []
  [n3]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[AuxKernels]
  [normal_dirn_x]
    type = PorousFlowElementNormal
    variable = n1
    component = x
  []
  [normal_dirn_y]
    type = PorousFlowElementNormal
    variable = n2
    component = y
  []
  [normal_dirn_z]
    type = PorousFlowElementNormal
    variable = n3
    component = z
  []
[]
```

You can also see more examples [here](https://mooseframework.inl.gov/source/auxkernels/PorousFlowElementNormal.html).
Then, you will have each normal as an elemental variable in your (`exodus`) output.

**Reply by TLWise** [reply]
@Ali1990dashti, I am using the Combined App, modeling channel flow.

**Reply by TLWise** [reply]
when I included the code into my file.i I received the following error: 
*** ERROR ***
A 'PorousFlowElementNormal' is not a registered object.

If you are trying to find this object in a dynamically loaded library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.

**Reply by Ali1990dashti** [reply]
The error you faced is discussed [here](https://github.com/idaholab/moose/discussions/18882).

**Reply by TLWise** [reply]
After adding APPs to the make file of Parent(crab) APP, I had to $ make clobberall, followed by a make of the parent APP. I no longer get an error about 'PorousFlowElementNormal' is not a registered object. I do no however have this error:

*** ERROR ***
The parameter vectors constant_names and constant_values must have equal length.

**Reply by GiudGiud** [reply]
Hello

This is coming from one of your Parsed objects. It should say which one in the error message

And the message says what the problem is exactly

Guillaume

**Reply by TLWise** [reply]
Here is how I'm attempting to implement it:

[AuxVariables]
  [vel_x]
    order = FIRST
  []
  [vel_y]
    order = FIRST
  []
  [nx]
    order = CONSTANT
    family = MONOMIAL
  []
  [ny]
    order = CONSTANT
    family = MONOMIAL
  []
  [nz]
    order = CONSTANT
    family = MONOMIAL
  []  
[]

[AuxKernels]
  [vel_x]
    type = VectorVariableComponentAux
    variable = vel_x
    vector_variable = velocity
    component = 'x'
  []
  [vel_y]
    type = VectorVariableComponentAux
    variable = vel_y
    vector_variable = velocity
    component = 'y'
  []
  [nx]
    type = PorousFlowElementNormal
    boundary = 'CylWall'    
    variable = nx
    component = x
    1D_perp = '0 1 0'   
  []
  [ny]
    type = PorousFlowElementNormal
    boundary = 'CylWall'      
    variable = ny
    component = y
    1D_perp = '0 1 0'    
  []
  [nz]
    type = PorousFlowElementNormal
    boundary = 'CylWall'      
    variable = nz
    component = z
    1D_perp = '0 1 0'    
  []  
[]

[Postprocessors]
  [drag_force]
    constant_names = 'nx'
    # #    constant_expressions = '${rho} ${fparse 2/3*inlet_velocity} ${fparse 2*circle_radius}'
    type = ParsedPostprocessor
    function = 'mu_diff_vel_x - normal_x_pressure * nx'
    pp_names = 'mu_diff_vel_x normal_x_pressure'
    execute_on = 'INITIAL TIMESTEP_END'
  []
[]

**Reply by GiudGiud** [reply]
The ParsedPostprocessor syntax is the problem.

You need to define the symbols and the names for every constant 

So these two parameters must match

**Reply by TLWise** [reply]
Will the parsedpostprocessor work at all for these combination of auxvariables and postprocessors?

**Reply by GiudGiud** [reply]
The postprocessors yes.

The auxvariable no. These cannot be inputs as we do not know which value to use (they are fields, there are many values)

Use a postprocessor to perform the desired reduction operation (average?) on the auxiliary variable first
